README.md
# pinary

------
![Node.js CI](https://github.com/eviltik/pinary/workflows/Node.js%20CI/badge.svg)
[![npm version](https://badge.fury.io/js/pinary.svg)](https://badge.fury.io/js/pinary)
[![MIT Licence](https://badges.frapsoft.com/os/mit/mit.svg?v=103)](https://opensource.org/licenses/mit-license.php)
[![Dependency Status](https://david-dm.org/eviltik/pinary.svg)](https://david-dm.org/eviltik/pinary)


## Introduction

Yet another RPC client and server, with minimalistic publish/subscribe implementation.

* Server
  * TCP or TLS, as you want (TLS require certificates, see https://github.com/jfromaniello/selfsigned )
  * Persistant connection
  * Binary frames (fast)
  * Optional ZLIB compression
  * Minimalistic JSON Schema implementation (input integrity)
  * Handle Max Clients
  * Basic publish/subscribe support

* Client
  * Automatic reconnect
  * Internal RPC calls queue
    * Allow lazy client connect
    * Store calls when not connected and play calls when reconnect


## Install

```
npm install pinary
```

## Server

### Server: instantiation

```
const PinaryServer = require('pinary').server;
const server = new PinaryServer();

// with options (see below)
// const server = new PinaryServer({port:64000});

```

| Option                | Default                       |  Notes  
|-----------------------|-------------------------------|----------------
| useTLS                | false                         | Use clear TCP or TLS    
| useZLIB               | false                         | Use ZLIB compression
| maxClients (1)        | 10                            | Maximum number of simultaneous TCP connections
| timeoutData           | 1000                          | Delay before socket close if no data sent, in milliseconds
| host                  | 0.0.0.0                       | Listening IP/host
| port                  | 65000 for TCP, 65001 for TLS  | Listening port
| key                   | null                          | TLS: private key
| cert                  | null                          | TLS: public key
| ca                    | null                          | TLS: certificate authority (string of array of string)
| secureProtocol        | TLSv1_2_method                | TLS: cipher
| rejectUnauthorized    | false                         | TLS: allow self signed certificates, or not

(1) under the hood, a "client" is in fact 2 sockets, one for writing, one for reading.


## Client

### Client: instantiation
```
const PinaryClient = require('pinary').client;
const client = new PinaryClient(); // auto connect

// with a TCP url
// const client = new PinaryClient('tcp://localhost:64000',[options]);

// with a TLS url
// const client = new PinaryClient('tls://localhost:64000',[options]);
```
| Option                                | Default                       | Note |      
|---------------------------------------|-------------------------------|------|
| reconnectInterval                     | 500                           | milliseconds   |
| queueSize                             | 100                           | store rpc calls limit when not connected/disconnected |

### Client: trigger a method

```
// using callback
client.rpc('myMethod', (err, result) => {
    if (err) throw err;
    console.log(result);
});

// using async/await
async function letsgo() {
    let result;
    try {
        result = await client.rpcPromise('myMethod');
    } catch(e) {
        // something wrong
    }
    console.log(result);
}
```

Note: if not yet connected or while the client is trying to reconnect,
RPC calls are stored in a queue and played when client is connected.

#### Client: events

| event name            | arguments                     |  Notes  
|-----------------------|-------------------------------|----------------
| connected             | retryCount                    | if retryCount = 0, first connection, else reconnection  
| disconnected          |                               |
| error                 | Error                         |


## Publish/Subscribe (PUBSUB)

### Client to clients
```
const Server = require('pinary').server;
const Client = require('pinary').client;

const server = new Server();
const client1 = new Client();
const client2 = new Client();

const channel = '/myChannel';

server.start();

client1.subscribe('/bla', (data) => {
    console.log(data);
    process.exit();
});

client2.publish('/bla', { foo:'bar' });

```

#### Server to clients
```
const Server = require('pinary').server;
const Client = require('pinary').client;

const server = new Server();
const client = new Client();

const channel = '/myChannel';

server.start();

client.subscribe(channel, (data) => {
    console.log(data);
    process.exit();
});

server.publish(channel, { foo:'bar' });
```


The actual implementation is minimalistic:
* a channel is considered as an ID, you cannot use wildcards like redis or faye


## TODO
* finish doc
  * events emitted (server)
  * server methods registration (see test/tests/102.methodExist.js, or examples/ for moment)


CHANGELOG.md
# Changelog

## [1.3.8](https://github.com/eviltik/pinary/) - 2021-01-25

## [1.3.7](https://github.com/eviltik/pinary/) - 2019-10-21

## [1.2.5](https://github.com/eviltik/pinary/) - 2019-06-14

## [1.2.4](https://github.com/eviltik/pinary/) - 2019-05-08

### First release


src/server/index.js
const methods = require('./methods/');
const net = require('net');
const tls = require('tls');
const missive = require('missive');
const debug = require('debug')('pinary:server');
const merge = require('deepmerge');
const handler = require('./methods/handler');
const async = require('async');
const attributes = require('../attributes');
const frame = require('../frame');

const DEFAULT_OPTIONS = {
    useTLS:false,
    useZLIB: false,
    maxClients:10,
    host:'0.0.0.0',
    port:65000,
    key:null,
    cert:null,
    ca:null, // should be an array of ca
    secureProtocol: 'TLSv1_2_method',
    rejectUnauthorized: false,
    timeoutData:1000
};

const maxSIZE = 1024 * 1024 * 10; // 10 Mo

function Server(options) {

    options = merge(DEFAULT_OPTIONS, options||{});

    let server;
    let serverSubscribedChannels = {};
    let clientsSubscribedChannel = {};

    function RPCCall(task) {

        methods.exec(
            task.data.m,
            task.data.p,
            { server, session:task.socket },
            (err, result) => {

                if (err) {

                    if (err instanceof Error) {
                        task.encoder.write(frame.error(task.data.id, err.message));
                        debug(`${task.socket.id}: ${err.stack}`);
                        return;
                    }

                    task.encoder.write(frame.error(task.data.id, err));
                    debug(`${task.socket.id}: ${err.message}`);
                    return;
                }

                task.encoder.write(frame.result(task.data.id, result));
            }
        );
    }

    function removeClient(socketId) {
        delete server.clients[socketId];
        delete server.clientsReader[socketId];
        async.mapValues(clientsSubscribedChannel, (clientsId, channel, next) => {
            const idx = clientsId.indexOf(socketId);
            if (idx>=0) {
                clientsSubscribedChannel[channel].splice(idx, 1);
            }
            next();
        });
    }

    function onServerError(err) {
        debug(err);
    }

    function onConnect(socket) {
        socket.setNoDelay(true);
        server.emit('connect');

        socket.killer = setTimeout(() => {
            // kill the socket if nothing append on it
            socket.end();
        }, options.timeoutData);


        socket.id = `${socket.remoteAddress}:${socket.remotePort}`;

        socket.on('end', () => {
            debug(`${socket.id}: client socket end`);
            server.clients[socket.id].socket.destroy();
            removeClient(socket.id);
        });

        socket.on('error', err => {
            debug(`${socket.id}: ${err.message}`);
            removeClient(socket.id);
        });

        socket.once('data', () => {
            clearTimeout(socket.killer);
        });

        debug(`${socket.id}: client connected`);

        const encoder = missive.encode({ maxSize: maxSIZE, deflate: options.useZLIB });
        const decoder = missive.parse({ maxSize: maxSIZE, inflate: options.useZLIB });

        decoder.on('message', data => {

            const mid = data[attributes.id];
            const mmethod = data[attributes.method];
            const mparams = data[attributes.params];
            const mdata = data[attributes.data];
            const mchannel = data[attributes.channel];

            if (server._connections>(options.maxClients*2)) {
                debug(`${socket.id}: refusing connection, number of connection: ${server._connections-1}, allowed: ${options.maxClients*2}`);
                encoder.write(frame.error(mid, 'MAX_CLIENT_REACHED'));
                socket.end();
                return;
            }

            if (!mmethod) {
                debug(`${socket.id}: missing method in the payload`);
                encoder.write(frame.error(mid, 'MISSING_METHOD_ATTRIBUTE'));
                return;
            }

            if (mmethod === attributes.publish) {
                debug(`${socket.id}: pubsub: received data from channel ${mchannel}`);
                if (serverSubscribedChannels[mchannel]) {
                    serverSubscribedChannels[mchannel](mdata);
                }
                publish(mchannel, mdata);
                return;
            }

            if (mmethod === attributes.subscribe) {
                if (!clientsSubscribedChannel[mchannel]) {
                    clientsSubscribedChannel[mchannel] = [];
                }
                clientsSubscribedChannel[mchannel].push(socket.id);
                return;
            }

            if (!methods.exists(mmethod)) {
                debug(`${socket.id}: unknow method ${mmethod}`);
                encoder.write(frame.error(mid, `UNKNOW_METHOD ${mmethod}`));
                return;
            }

            if (mparams) {
                debug(`${socket.id}: method ${mmethod}: exec with params ${JSON.stringify(mparams)}`);
            } else {
                debug(`${socket.id}: method ${mmethod}: exec without params`);
            }

            RPCCall({ data, socket, encoder });
        });

        decoder.on('error', err => {
            debug(err);
        });

        socket.pipe(decoder);
        encoder.pipe(socket);

        server.clients[socket.id] = { socket, encoder };

    }

    function start(callback) {

        debug(`server starting ... (${options.host}:${options.port} maxClients ${options.maxClients*2})`);

        methods.initialize();

        function _onServerListen(err) {

            if (err) {
                debug(err);
                throw new Error(err);
            }

            debug('server started');
            callback && callback();
        }

        if (options.useTLS) {

            server = tls.createServer({
                key:options.key,
                cert:options.cert,
                ca:options.ca,
                secureProtocol:options.secureProtocol,
                rejectUnauthorized: options.rejectUnauthorized
            });

            server.on('tlsClientError', (err) => {
                debug(err.message);
            });

            server.on('secureConnection', onConnect);

        } else {

            server = net.createServer();
            server.on('connection', onConnect);

        }

        server.clients = {};
        server.clientsReader = {};

        server.on('listening', _onServerListen);
        server.on('error', onServerError);

        server.listen(options.port, options.host);

    }

    function stop(callback) {
        let closed = 0;
        try {
            let id;
            for (id in server.clients) {
                debug(`server stopping: ending socket ${id}`);
                server.clients[id].encoder.write(JSON.stringify({ error:'SERVER_SHUTDOWN' }));
                server.clients[id].socket.end();
                server.clients[id].socket.destroy();
                removeClient(id);
                closed++;
            }
        } catch(e) {
            debug(e);
        }

        serverSubscribedChannels = {};
        clientsSubscribedChannel = {};

        if (closed) {
            debug(`${closed} socket(s) has been closed, ${closed/2} client disconnected`);
        }

        server.close(err => {
            if (err) {
                debug(err);
            }

            debug('server closed');

            if (callback) {
                callback();
            }
        });
    }

    function registerMethod(descriptor, fn) {
        methods.register(new handler.Method(descriptor, fn));
    }

    function on(eventName, fn) {
        server.on(eventName, fn);
    }

    function publish(channel, data) {

        if (!clientsSubscribedChannel[channel] || !clientsSubscribedChannel[channel].length) {
            debug(`pubsub: dispatch: no subscribers for channel ${channel} (server only ?)`);
            return;
        }

        let send = 0;

        async.mapValues(clientsSubscribedChannel[channel], (writerId, index, next) => {
            if (!server.clients[writerId]) {
                debug(`pubsub: dispatch error: client ${writerId} not found`);
                return;
            }

            if (!server.clients[writerId].socket) {
                debug(`pubsub: dispatch error: client ${writerId} don't have underlying socket object`);
                return;
            }

            if (!server.clients[writerId].socket.socketReader) {
                debug(`pubsub: dispatch error: client ${writerId} has no socketReader attached`);
                return;
            }

            const readerId = server.clients[writerId].socket.socketReader.socket.id;
            const reader = server.clientsReader[readerId];
            try {
                debug(`pubsub: dispatching: message sent to client ${readerId}`);
                reader.write(frame.publish(channel, data));
                send++;
            } catch(e) {
                debug(e);
            }
            next();
        }, () => {
            //debug(`pubsub: dispatching: sent to ${send} clients`);
        });
    }

    function subscribe(channel, callback) {
        debug(`pubsub: server subscribe to ${channel}`);
        serverSubscribedChannels[channel] = callback;
    }

    return {
        start,
        stop,
        registerMethod,
        on,
        publish,
        subscribe
    };
}

module.exports = Server;
